stan/math/opencl/matrix_cl.hpp

```diff
@@ -322,7 +319,7 @@ class matrix_cl<T, enable_if_arithmetic<T>> {
       buffer_cl_ = cl::Buffer(ctx, CL_MEM_READ_WRITE, sizeof(T) * A.size());
       cl::Event transfer_event;
       queue.enqueueWriteBuffer(buffer_cl_, CL_FALSE, 0, sizeof(T) * A.size(),
-                               A.data(), nullptr, &transfer_event);
+                               A.eval().data(), nullptr, &transfer_event);
```

Comment:
Indeed! I think we always want to right now right? Else the bug you saw earlier. Maybe some of these thing should go in a separate PR

Reply:
These changes to `matrix_cl` should definitely go in a separate PR. And I think we only want to evaluate expressions that are not `Eigen::Ref` or `Eigen::Map`.

Reply:
> Calling .eval() on Eigen::Ref or Eigen::Map will make a copy.

I'm looking at the docs, If we map to a `Matrix` type won't it call `DenseBase::eval()` which doesn't make a copy?

> Notice that in the case of a plain matrix or vector (not an expression) this function just returns a const reference, in order to avoid a useless copy.

Or is a Map/Ref considered an expression?

https://eigen.tuxfamily.org/dox/classEigen_1_1DenseBase.html#aa73e57a2f0f7cfcb4ad4d55ea0b6414b

Reply:
Since this is templated now we won't get `Eigen::Ref` anyway. I guess supporting `Eigen::Map` without copies can wait.

Reply:
Did you see my comment above? Doesn't Map on a Matrix use DenseBase::eval() which does not make a copy when calling `.eval()`?

Reply:
eval on Matrix is a no-op. eval on Map will make a Matrix from it - a copy.